File: packages/vela-sdk-ts/src/mcp/auto-advance.ts
/**
 * Auto-advance loop logic for workflow execution.
 *
 * Elicit -> commit -> advance loop. Stops at steps that require
 * agent work: EXECUTE, DIALOG, WORKFLOW, or incomplete captures.
 */

import type {
  AdvanceResult,
  WorkflowRunState,
} from "../engine/types.js";
import type { IWorkflowEngine } from "../engine/workflow-engine.js";
import type { ResourceResolver } from "../engine/prompt-builder.js";
import type { Locale } from "../locale/locale.js";
import { getLocale } from "../locale/locale.js";
import type {
  AnyStepDefinition,
  WorkflowDefinition,
} from "../schemas/workflow.js";
import type { WorkflowStore } from "../storage/store.js";
import type { McpContext } from "./mcp-server.js";
import { ElicitationService } from "./elicitation.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isAutomationMode(run: WorkflowRunState): boolean {
  const val = run.params["automation_mode"];
  return val === true || val === "true" || val === 1;
}

// ---------------------------------------------------------------------------
// stepCapturesComplete
// ---------------------------------------------------------------------------

export function stepCapturesComplete(
  stepDef: AnyStepDefinition,
  run: WorkflowRunState,
): boolean {
  const elicitable = stepDef.capture.filter((cap) => cap.elicit !== "never");
  if (elicitable.length === 0) {
    return true;
  }
  return elicitable.every((cap) => cap.key in run.stateData);
}

// ---------------------------------------------------------------------------
// elicitStepCaptures
// ---------------------------------------------------------------------------

export async function elicitStepCaptures(
  ctx: McpContext,
  engine: IWorkflowEngine,
  wfDef: WorkflowDefinition,
  run: WorkflowRunState,
  store: WorkflowStore,
  locale?: Locale,
): Promise<WorkflowRunState> {
  const loc = locale ?? getLocale();

  const stepDef = engine.getStep(wfDef, run.currentStep ?? "");
  if (!stepDef || !stepDef.capture || stepDef.capture.length === 0) {
    return run;
  }

  // Dialog steps handle captures after all phases complete
  if (stepDef.type === "dialog") {
    return run;
  }

  let currentRun = run;

  for (const cap of stepDef.capture) {
    if (cap.elicit === "never") {
      continue;
    }
    // In automation mode: only elicit required captures, skip the rest
    if (isAutomationMode(currentRun) && !cap.required) {
      continue;
    }

    const existingValue = currentRun.stateData[cap.key];
    if (cap.elicit === "if_missing" && existingValue !== undefined && existingValue !== null) {
      continue;
    }

    const schema = ElicitationService.buildRequestedSchema(cap);
    let message = ElicitationService.buildMessage(cap);
    if (existingValue !== undefined && existingValue !== null) {
      message += loc.currentValueHint.replace(
        "{existing_value}",
        String(existingValue),
      );
    }

    let elicitResult;
    try {
      elicitResult = await ctx.elicit(message, schema);
    } catch {
      ctx.log.debug("elicitation.not_supported", { key: cap.key });
      return currentRun;
    }

    const processed = ElicitationService.processResult(cap, elicitResult);
    if (processed) {
      const [key, value] = processed;
      currentRun = await store.updateStep(currentRun.id, null, {
        stateData: { [key]: value },
      });
      await store.commit();
    }
  }

  return currentRun;
}

// ---------------------------------------------------------------------------
// autoAdvanceLoop
// ---------------------------------------------------------------------------

export async function autoAdvanceLoop(
  ctx: McpContext,
  engine: IWorkflowEngine,
  wfDef: WorkflowDefinition,
  result: AdvanceResult,
  store: WorkflowStore,
  resourceResolver?: ResourceResolver,
  locale?: Locale,
): Promise<AdvanceResult> {
  let current = result;

  while (!current.completed && current.run.currentStep && ctx) {
    const stepDef = engine.getStep(wfDef, current.run.currentStep ?? "");
    if (!stepDef) {
      break;
    }

    // Execute steps always need agent work
    if (stepDef.type === "execute") {
      break;
    }

    // Dialog steps need interactive multi-phase agent work
    if (stepDef.type === "dialog") {
      break;
    }

    // Workflow steps delegate to sub-workflows (handled by orchestrator)
    if (stepDef.type === "workflow") {
      break;
    }

    // Steps without elicitable captures need agent output
    const hasElicitable = stepDef.capture
      ? stepDef.capture.some((c) => c.elicit !== "never")
      : false;

    if (!hasElicitable) {
      const interactiveTypes = new Set(["confirm", "choice", "freeform"]);
      if (!interactiveTypes.has(stepDef.type)) {
        // If this is the final step, auto-advance to complete the workflow
        // We check by trying to advance — the engine resolves next step internally
        current = await engine.advance(current.run, wfDef, {
          resourceResolver,
        });
        await store.commit();
      }
      break;
    }

    current.run = await elicitStepCaptures(
      ctx,
      engine,
      wfDef,
      current.run,
      store,
      locale,
    );

    if (stepCapturesComplete(stepDef, current.run)) {
      current = await engine.advance(current.run, wfDef, {
        resourceResolver,
      });
      await store.commit();
    } else {
      await store.commit();
      break;
    }
  }

  return current;
}

File: packages/vela-sdk/src/vela_sdk/fastmcp/auto_advance.py
"""Auto-advance loop logic for workflow execution."""

import json
from typing import Any, Callable, Optional

import structlog

from vela_sdk.engine.types import AdvanceResult, WorkflowRunState, WorkflowRunStatus
from vela_sdk.engine.workflow_engine import WorkflowEngine
from vela_sdk.fastmcp.elicitation import ElicitationService
from vela_sdk.fastmcp.locale import Locale, get_locale
from vela_sdk.schemas.resource import ResourceDefinition
from vela_sdk.schemas.workflow import StepType, WorkflowDefinition
from vela_sdk.storage.protocol import WorkflowStore

logger = structlog.get_logger()


def _is_automation_mode(run: WorkflowRunState) -> bool:
    """Return True if automation_mode param is set to a truthy value."""
    val = run.params.get("automation_mode")
    return val is True or val == "true" or val == 1


def step_captures_complete(step_def: Any, run: WorkflowRunState) -> bool:
    """Check if all elicitable captures for a step are present in state."""
    elicitable = [cap for cap in step_def.capture if cap.elicit != "never"]
    if not elicitable:
        return True
    return all(cap.key in run.state_data for cap in elicitable)


async def elicit_step_captures(
    ctx: Any,
    engine: WorkflowEngine,
    wf_def: WorkflowDefinition,
    run: WorkflowRunState,
    store: WorkflowStore,
    locale: Optional[Locale] = None,
) -> WorkflowRunState:
    """Run elicitation loop for the current step's captures.

    Returns the (possibly updated) run state.
    """
    if locale is None:
        locale = get_locale()

    step_def = engine._get_step(wf_def, run.current_step)
    if not step_def or not step_def.capture:
        return run

    # Dialog steps handle captures after all phases complete
    if step_def.type == StepType.DIALOG:
        return run

    for cap in step_def.capture:
        if cap.elicit == "never":
            continue
        # In automation mode: only elicit required captures, skip the rest
        if _is_automation_mode(run) and not cap.required:
            continue
        # Re-read state_data from current run (may have been updated by previous capture)
        existing_value = run.state_data.get(cap.key)
        if cap.elicit == "if_missing" and existing_value is not None:
            continue

        response_type = ElicitationService.build_response_type(cap)
        message = ElicitationService.build_message(cap)
        if existing_value is not None:
            message += locale.current_value_hint.format(existing_value=existing_value)

        try:
            elicit_result = await ctx.elicit(message, response_type)
        except Exception:
            logger.debug("elicitation.not_supported", key=cap.key)
            return run

        processed = ElicitationService.process_result(cap, elicit_result)
        if processed:
            key, value = processed
            run = await store.update_step(run.id, None, state_data={key: value})
            await store.commit()

    return run


async def auto_advance_loop(
    ctx: Any,
    engine: WorkflowEngine,
    wf_def: WorkflowDefinition,
    result: AdvanceResult,
    store: WorkflowStore,
    resource_resolver: Optional[Callable[[str], Optional[ResourceDefinition]]] = None,
    locale: Optional[Locale] = None,
) -> AdvanceResult:
    """Elicit -> commit -> advance loop. Runs until a step needs agent work.

    Stops at steps that require agent work:
    - Execute steps (agent must perform task + provide output)
    - Dialog steps (multi-phase interactive)
    - Steps with incomplete captures after elicitation
    - Choice/freeform steps without captures (need agent output for branching)
    """
    while not result.completed and result.run.current_step and ctx:
        step_def = engine._get_step(wf_def, result.run.current_step)
        if not step_def:
            break

        # Execute steps always need agent work
        if step_def.type == StepType.EXECUTE:
            break

        # Dialog steps need interactive multi-phase agent work
        if step_def.type == StepType.DIALOG:
            break

        # Workflow steps delegate to sub-workflows (handled by orchestrator)
        if step_def.type == StepType.WORKFLOW:
            break

        # Steps without elicitable captures need agent output
        has_elicitable = any(c.elicit != "never" for c in step_def.capture) if step_def.capture else False
        if not has_elicitable:
            interactive_types = {StepType.CONFIRM, StepType.CHOICE, StepType.FREEFORM}
            if step_def.type not in interactive_types:
                # If this is the final step, auto-advance to complete the workflow
                next_step_id = engine._resolve_next(step_def, None, wf_def)
                if next_step_id is None:
                    result = await engine.advance(
                        result.run, wf_def, resource_resolver=resource_resolver
                    )
                    await store.commit()
            break

        result.run = await elicit_step_captures(ctx, engine, wf_def, result.run, store, locale=locale)
        if step_captures_complete(step_def, result.run):
            result = await engine.advance(
                result.run, wf_def, resource_resolver=resource_resolver
            )
            await store.commit()
        else:
            await store.commit()
            break
    return result
